include/lsst/meas/extensions/shapeHSM/HsmMomentsControl.h

```diff
@@ -1,46 +1,154 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
 #ifndef LSST_MEAS_EXTENSIONS_SHAPEHSM_HsmMomentsControl_h_INCLUDED
 #define LSST_MEAS_EXTENSIONS_SHAPEHSM_HsmMomentsControl_h_INCLUDED
 
-#include "lsst/meas/algorithms/ShapeControl.h"
+#include "lsst/pex/config.h"
+#include "lsst/meas/base/Algorithm.h"
+#include "lsst/meas/base/ShapeUtilities.h"
+#include "lsst/meas/base/CentroidUtilities.h"
+#include "lsst/meas/base/FlagHandler.h"
+#include "lsst/meas/base/InputUtilities.h"
 
 namespace lsst { namespace meas { namespace extensions { namespace shapeHSM {
 
-class HsmSourceMomentsControl : public algorithms::ShapeControl {
+// This is here just as a parent class for the following
+class HsmMomentsControl {
+};
+
+class HsmSourceMomentsControl : public HsmMomentsControl {
 public:
     LSST_CONTROL_FIELD(badMaskPlanes, std::vector<std::string>, "Mask planes used to reject bad pixels.");
 
-    HsmSourceMomentsControl() : ShapeControl("shape.hsm.moments") {}
-
     HsmSourceMomentsControl(HsmSourceMomentsControl const & other) :
-        algorithms::ShapeControl(other), badMaskPlanes(other.badMaskPlanes) {}
+        badMaskPlanes(other.badMaskPlanes) {}
 
-    HsmSourceMomentsControl(std::string const & name) : algorithms::ShapeControl(name) {
+    HsmSourceMomentsControl() {
         badMaskPlanes.push_back("BAD");
         badMaskPlanes.push_back("SAT");
         badMaskPlanes.push_back("INTRP");
     }
-private:
-    virtual PTR(algorithms::AlgorithmControl) _clone() const;
-    virtual PTR(algorithms::Algorithm) _makeAlgorithm(
-        afw::table::Schema & schema, PTR(daf::base::PropertyList) const & metadata
-    ) const;
 };
 
+class HsmPsfMomentsControl : public HsmMomentsControl {
+public:
+    HsmPsfMomentsControl() {}
+
+    HsmPsfMomentsControl(HsmPsfMomentsControl const & other) {}
+};
 
-class HsmPsfMomentsControl : public algorithms::ShapeControl {
+/// Base class to measure HSM adaptive moments
+///
+/// Use this to consolidate common code for HsmSourceMoments and HsmPsfMoments
+class HsmMomentsAlgorithm : public base::SimpleAlgorithm {
 public:
-    HsmPsfMomentsControl() : ShapeControl("shape.hsm.psfMoments") {}
-    HsmPsfMomentsControl(std::string const & name) : algorithms::ShapeControl(name) {}
-    HsmPsfMomentsControl(HsmPsfMomentsControl const & other) : algorithms::ShapeControl(other) {}
+    /// A typedef to the Control object for this algorithm, defined above.
+    ///  The control object contains the configuration parameters for this algorithm.
+    typedef HsmMomentsControl Control;
+
+    /// @brief Initialize with standard field names and customized documentation.
+    HsmMomentsAlgorithm(Control const & ctrl, std::string const & name, afw::table::Schema & schema, char const* doc) :
```

Comment:
Also, this line is too long (but I think removing the control argument will fix that).